File: CTDash/managers/contrast_manager.py
# [LATER] — depends on patient states being stable
# ContrastManager — tracks oral contrast timers and injector fill status.
# Fires ContrastReady state change when oral contrast timer completes.
# Applies a penalty if IV scan starts before injector is filled.

from classes.patient import PatientState
from config import ORAL_CONTRAST_WAIT, INJECTOR_FILL_TIME


class ContrastManager:
    def __init__(self):
        self._oral_patients  = {}    # patient_id -> Patient (oral contrast waiting)
        self._injector_ready = {}    # patient_id -> bool

    def start_oral_contrast(self, patient):
        """Begin oral contrast wait timer for a patient."""
        patient.contrast_timer = ORAL_CONTRAST_WAIT
        patient.state          = PatientState.CONTRAST_ORDERED
        self._oral_patients[patient.patient_id] = patient

    def fill_injector(self, patient):
        """Signal that the IV injector has been filled for a patient."""
        self._injector_ready[patient.patient_id] = True
        patient.state = PatientState.INJECTOR_READY

    def tick(self):
        """Count down oral contrast timers. Fire ContrastReady when timer hits zero."""
        ready = []
        for pid, patient in self._oral_patients.items():
            patient.contrast_timer -= 1
            if patient.contrast_timer <= 0:
                patient.contrast_timer = 0
                patient.state          = PatientState.CONTRAST_READY
                ready.append(pid)

        for pid in ready:
            del self._oral_patients[pid]

File: CTDash_Unity/Assets/Scripts/Managers/ContrastManager.cs
// ContrastManager — tracks oral contrast timers and injector fill status.
// Fires ContrastReady state change when oral contrast timer completes.

using System.Collections.Generic;

namespace CTDash
{
    public class ContrastManager
    {
        private readonly Dictionary<string, Patient> _oralPatients  = new Dictionary<string, Patient>();
        private readonly HashSet<string>             _injectorReady = new HashSet<string>();

        public void StartOralContrast(Patient patient)
        {
            patient.ContrastTimer = GameConfig.ORAL_CONTRAST_WAIT;
            patient.State         = PatientState.ContrastOrdered;
            _oralPatients[patient.PatientId] = patient;
        }

        public void FillInjector(Patient patient)
        {
            _injectorReady.Add(patient.PatientId);
            patient.State = PatientState.InjectorReady;
        }

        public bool IsInjectorReady(string patientId) => _injectorReady.Contains(patientId);

        public void Tick()
        {
            var ready = new List<string>();
            foreach (var kvp in _oralPatients)
            {
                kvp.Value.ContrastTimer -= 1;
                if (kvp.Value.ContrastTimer <= 0)
                {
                    kvp.Value.ContrastTimer = 0;
                    kvp.Value.State         = PatientState.ContrastReady;
                    ready.Add(kvp.Key);
                }
            }
            foreach (var pid in ready)
                _oralPatients.Remove(pid);
        }
    }
}
